rust/gui-client/src-tauri/src/client/network_changes/windows.rs

```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+            let mut wait_handle = HANDLE(0isize);
+
+            // The docs say that `RegisterWaitForSingleObject` uses "a worker thread" from
+            // "the thread pool".
+            // Raymond Chen, who is an authority on Windows internals, says RegisterWaitForSingleObject
+            // does multiple waits on the same worker thread internally:
+            // <https://devblogs.microsoft.com/oldnewthing/20081117-00/?p=20183>
+            // There is a different function in the CLR called `RegisterWaitForSingleObject`,
+            // so he might be talking about that.
+            // If he's not, then we're invisibly tying up two worker threads. Can't help it.
+
+            // SAFETY: It's complicated.
+            // The callback here can cause a lot of problems. We pin the `Notify` object.
+            // We don't use an `Arc` because `Notify` already has only `&self` methods, and
+            // the callback has no way to free an `Arc`.
+            // When we call `UnregisterWaitEx` later, we wait for all callbacks to finish running
+            // before we drop the `Notify`, to avoid a dangling pointer.
+            unsafe {
+                RegisterWaitForSingleObject(
+                    &mut wait_handle,
+                    event,
+                    Some(callback),
+                    Some(notify_ptr as *const _),
+                    INFINITE,
+                    WT_EXECUTEINWAITTHREAD,
+                )
+            }?;
+
+            let mut that = Self {
+                key,
+                notify,
+                wait_handle: Some(wait_handle),
+                event: Some(event),
+            };
+            that.register_callback()?;
+
+            Ok(that)
+        }
+
+        pub(crate) fn close(&mut self) -> Result<()> {
+            if let Some(wait_handle) = self.wait_handle.take() {
+                unsafe { UnregisterWaitEx(wait_handle, INVALID_HANDLE_VALUE) }?;
+            }
+            if let Some(event) = self.event.take() {
+                unsafe { CloseHandle(event) }?;
+            }
+
+            Ok(())
+        }
+
+        // I'm not sure if this needs to be mut, I'm being cautious here.
+        // There is no reason in practice you'd ever want to call it from two
+        // threads at once.
+        //
+        // Cancel-safety: Probably. `tokio::sync::Notify` sounds cancel-safe.
+        pub(crate) async fn notified(&mut self) -> Result<()> {
+            // Order matters here
+            // We want to return to the caller and let them check the DNS
+            // while the callback is registered.
+            self.notify.notified().await;
+            self.register_callback()
+                .context("`register_callback` failed")?;
+
+            Ok(())
+        }
+
+        // Be careful with this, if you register twice before the callback fires,
+        // it will probably leak something. Check the MS docs for `RegNotifyChangeKeyValue`.
```

Comment:
This should be an `unsafe` function

Reply:
Because of the constraint mentioned in that comment?
🤔 Since it's all `&mut self`, what if I enforce it at runtime with a flag `callback_armed`?

Reply:
I mean, it's okay to mark this as unsafe, specially since we use it only internally, the methods using it can make sure that they are enforcing the constaints.

Reply:
I don't think it's memory-unsafe, I was thinking of this sentence from the docs https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regnotifychangekeyvalue#remarks

> Each time a process calls RegNotifyChangeKeyValue with the same set of parameters, it establishes another wait operation, creating a resource leak. Therefore, check that you are not calling RegNotifyChangeKeyValue with the same parameters until the previous wait operation has completed.

I might write a unit test for some of the registry stuff, there's no equivalent `Unregister` function, there's just closing the event handle. Even though it's only closed at the end of the process, it would be nice to know that editing the key won't break anything.